### ISSUE TITLE
Doesn't support PHPSpec 3.0

### ISSUE THREAD

**Issue by Dragonrun1** [henrikbjorn/PhpSpecCodeCoverageExtension#54]
Know they just released it but thought I'd remind you with new issue ;)

**Comment by kwisatz** [comment]
The dev-master branch is supposed to work with phpspec 3 but the phpspec 3.x configuration file syntax has changed and some interfaces have been renamed (see http://www.phpspec.net/en/stable/manual/upgrading-to-phpspec-3.html#upgrading-for-extension-authors)

Even though I've been able to get it to kinda run with a few adjustment, the services do not really seem to register with the service container.

So, the current dev-master is not yet working with phpspec 3.0

**Comment by Dragonrun1** [comment]
Yeah kind of played with it here locally to see if I could figure it out and just do a PR but didn't have any luck myself so started the issue to let someone else try figuring it out that knew the code better ;)

**Comment by henrikbjorn** [comment]
Feel free to update so we can get it working and cut a new major release.

**Comment by kwisatz** [comment]
Do you guys know of any other extension that is already working with phpspec 3.0? 

I adapted the code to agree with the new interface, but as I said, it didn't seem to actually register the listener. 

Also, when manually requesting the code_coverage class by its tag from the container, I would get an exception by the code coverage class, saying that there was no Coverage driver.

I can fork and commit the changes I did, @Dragonrun1 could check if maybe he had any additional things to be changed, but my code alone is currently not working and I don't really know where to look for more hints. Going back and forth between PHPSpec proper and Symfony\Console\Application is not fun and the phpspec extensions documentation is… ah… very lean.

**Comment by kwisatz** [comment]
I pushed what I have to https://github.com/tentwentyfour/PhpSpecCodeCoverageExtension/commit/f4c3056b3fcb98302948d370719bc4d5893b195b

I changed the namespace since I don't think it should be in the PhpSpec namespace, but maybe that's just a personal preference.

However, none of the lambda functions "defined" in the load() function are ever executed. I know that CodeCoverageExtension::load() is executed, but when PhpSpec has finished, 
```php
$container->getByTag('event_dispatcher.listeners.code_coverage') 
```
returns an empty array and code_coverage is not contained in 
```php
$container->getByTag('event_dispatcher.listeners') 
```

**Comment by kwisatz** [comment]
@henrikbjorn I have a few questions regarding some of your design choices, can I ask them here or do you prefer another form of communication for that?

**Comment by henrikbjorn** [comment]
Sure just ask here.

**Comment by henrikbjorn** [comment]
@kwisatz the reason for the container not returning our listener when you call `getByTag` is because it is not tagged. 

https://github.com/tentwentyfour/PhpSpecCodeCoverageExtension/blob/f4c3056b3fcb98302948d370719bc4d5893b195b/src/CodeCoverageExtension.php#L97 should have a 3rd argument which is the tag name `event_dispatcher.listeners` https://github.com/phpspec/phpspec/blob/d2a3c6033c2b028fb248f70cba07056038118619/src/PhpSpec/ServiceContainer/IndexedServiceContainer.php#L106

**Comment by kwisatz** [comment]
@henrikbjorn OK, so regarding the design choices:

1. In the listener, you have a setIO() method instead of passing ConsoleIO directly into the constructor. Is there a specific reason for this? I compared with core Listeners, which all inject the console IO when instantiating the class
2. You're also using array_map to apply the options. Any particular reason you chose not to use array_walk here?

**Comment by kwisatz** [comment]
@henrikbjorn OK, adding the tag, now I got it working. However, it seems to no longer be working with phpdbg. CodeCoverage complains that "No Code Coverage Driver available" when running phpspec through phpdbg.

**Comment by Dragonrun1** [comment]
So little late getting back to this but sounds like he spotted the problem and you'd already made the other changes I'd been trying. The one other thing I notice is you have in the README.md the setting being under extension.ExtName but the defines seem to still be looking at the old top level location instead. That probably needs fixed. Moving them I think is a good idea btw I've always thought they should have been there to start with.

**Comment by kwisatz** [comment]
@Dragonrun1 actually, I prefer the new location for the options and if nothing is found, I still accept the old way of defining them. I don't know if that should be removed, but currently, it doesn't hurt.

**Comment by Dragonrun1** [comment]
Do a composer dumpautoload and make sure you use phpdbg -qrr. I've found those fix a lot problems because of PHPSpec use of PSR-0 etc

**Comment by Dragonrun1** [comment]
Okay I must have missed where you check the new locations. Having it do both is a good idea for now and doc that the old location is going away in the future so people have chance to update.

**Comment by kwisatz** [comment]
@Dragonrun1 I still have the same behavior. I might be doing something wrong, but when using php5dbg -qrr vendor/bin/phpspec -v run, I quite simply get the following exception:

```
  [SebastianBergmann\CodeCoverage\RuntimeException]  
  No code coverage driver available                  
```

I pushed my newest changes to the repo at https://github.com/tentwentyfour/PhpSpecCodeCoverageExtension

@henrikbjorn is that something you'd accept as a PR or should we look into the questions above and the phpdbg issues first?

**Comment by henrikbjorn** [comment]
@kwisatz i will gladly accept it. People can use xdebug until phpdbg is fixed.

The phpdbg error seems to come from the Environment package it uses https://github.com/sebastianbergmann/environment/blob/master/src/Runtime.php#L190 which i guess returns false for your usage. Is it possible to you to debug why?

**Comment by henrikbjorn** [comment]
@kwisatz regarding design choices.

1. I don't remember why i do a setIO instead of injecting it through the constructor.
2. No specific reason behind it.

**Comment by kwisatz** [comment]
@henrikbjorn Now i know why you used array_map instead of array_walk ;) Updated the repo, will be looking into the phpdbg runtime error thing a little later.

**Comment by kwisatz** [comment]
OK, found the problem. Sebastian\Environment requires PHP7 for phpdbg.
It checks for the function phpdbg_start_oplog, which does not exist in phpdbg for PHP5.6.
See https://github.com/krakjoe/phpdbg/issues/150

**Comment by henrikbjorn** [comment]
@kwisatz Awesome detective work you have done. Thanks for your work on this, i have commented on the PR you opened and it is very small stuff that needs changing. Looking forward to merging it soon.

**Comment by kwisatz** [comment]
@henrikbjorn Thanks, will fix the changes and update the PR

**Comment by henrikbjorn** [comment]
I will cut a new release tomorrow :)!

**Comment by Dragonrun1** [comment]
Another way to fix problem with PHP 5.6 is just change required version of PHP in composer.json to >7.0 ;)

**Comment by henrikbjorn** [comment]
That would be a poor solution
lør. 30. jul. 2016 kl. 05.40 skrev Michael Cummings <
notifications@github.com>:

**Comment by Dragonrun1** [comment]
Probably but considering it's only a few months before it hits EOL it might be an idea in the near future. If everyone keeps support older versions as a community we'll always be stuck in the past.

**Comment by Dragonrun1** [comment]
BTW just tried out dev-master with one of my projects that is using PHPSpec 3.0 and it seems to be all working for me. Also need to update extenation example in README.md for new namespace but otherwise just waiting on new tagged version.

**Comment by Dragonrun1** [comment]
Just was updating by phpspec.yml file and trying with the settings under the extension name and that doesn't seem to work for me only in the old location. Do you want me to open another issue for that since my issue with it not working with PHPSpec 3.0 is fixed or is that just something that wasn't in the PR and you plan on doing another one?

**Comment by kwisatz** [comment]
@Dragonrun1 Here's my phpspec.yml, it works fine with the new syntax:

```yaml
extensions:
  PhpSpecCodeCoverage\CodeCoverageExtension:
    whitelist:
      - src
    format:
      - text
      - clover
    output:
      clover: build/clover.xml
```

Can you post yours so we can check what might be the problem with yours?

**Comment by Dragonrun1** [comment]
That makes sense I'll do that no sure why I didn't think of it.
